### ISSUE TITLE
FSDP reduce-overhead regression from CUDA allocator block ordering

### ISSUE THREAD

**Issue by rdspring1** [pytorch/pytorch#183451]
## Summary
PyTorch PR #178362 / commit 3e263a46d03bbd64637b0607fe4d0d3c7ca0fa17 regressed FSDP with `torch.compile(mode="reduce-overhead")` by changing CUDA allocator block ordering from address-based to allocation-counter-based. It happens in FSDP-only multi-GPU CUDA graph workloads, where unstable static input pointers trigger excessive cudagraph re-recording. Before `3e263a46`, 4xFSDP bf16 reduce-overhead had 40 re-records, 240 pointer changes and scaled from 237,167 tok/s at 1x to 760,795 tok/s at 4x (3.21x, 80.2% efficiency). With `3e263a46`, it regressed to 144 re-records, 1,588 pointer changes and scaled only from 226,767 tok/s to 598,963 tok/s (2.64x, 66.0% efficiency). This is not a permanent throughput ceiling. Peak throughput eventually returns, but only after a substantially longer warm-up period with repeated CUDA graph re-recording.

## Repro
[Github Gist - test_llama3_mlp_cudagraph_regression.py](https://gist.github.com/rdspring1/a4b818755156d14064a487437bd3af60)

## Training Loss Curve - 4xFSDP NVFP4 Llama3 8B

* Normally, throughput (token/sec) reaches high throughput within 5 steps. After this https://github.com/pytorch/pytorch/commit/3e263a46d03bbd64637b0607fe4d0d3c7ca0fa17, it can take ~2000 steps or 75M tokens.

<img width="1484" height="883" alt="Image" src="https://github.com/user-attachments/assets/5c458f27-1902-4a0a-aa4b-0e5337ca4d06" />

### AI Usage
* Codex for git bisect; Claude for analysis

### Versions

| Field | Prior commit - Good | Failing container |
|---|---|---|
| PyTorch git | 8c01604fcad5f08ffa39308e393e24ade29aa5eb | 3e263a46d03bbd64637b0607fe4d0d3c7ca0fa17 |
| CUDA | 13.3 | 13.3 |
| Python | 3.12.3 | 3.12.3 |
| GPU | 4x NVIDIA GB200 | 4x NVIDIA GB200 |
| Driver | 580.65.06 | 580.65.06 |
| Kernel | 6.14.0-1007-nvidia-64k | 6.11.0-1011-nvidia-64k |

cc @jerryzh168 @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng @zhaojuanmao @mrshenli @rohan-varma @chauhang @mori360 @ppwwyyxx

**Comment by eellison** [comment]
@rdspring1 i am going to submit a pr that would limit re-recordings for changed ptrs, so that they get just get copied. is that sufficient ?

**Comment by rdspring1** [comment]
Perhaps. I created a pytest suite to test any fix on. [test_llama3_mlp_cudagraph_regression.py](https://gist.github.com/rdspring1/87b07b516fc9b1bdd296d0454073888f#file-test_llama3_mlp_cudagraph_regression-py) measures the number of pointer changes and FSDP speedup.

**Comment by ngimel** [comment]
Can you dig into why ptrs are changing? https://github.com/pytorch/pytorch/pull/178362 aims to make allocation sequence deterministic whereas previously it was dictated by system returning addresses which was not deterministic, so I'd expect it to reduce pointer changes, not the other way around. I can hold off on landing #183591 as it might change the pointer order too. https://github.com/pytorch/pytorch/pull/183489 relanded #178362 with a smaller scope, for custom allocator pools only

**Comment by rdspring1** [comment]
Looks like https://github.com/pytorch/pytorch/pull/183591 fixes the issue. The address is included in the comparator, so within-segment ordering = address ordering = exactly the pre-#178362 stable behavior. Cudagraph trees converges.

  | Smoke log | Re-records | Static-pointer reasons | Individual pointer-change lines |
  |---|---:|---:|---:|
  | PR 183591 NVFP4 smoke | 72 | 72 | 4,632 |
  | Prior Good - https://github.com/pytorch/pytorch/commit/8c01604fcad5f08ffa39308e393e24ade29aa5eb | 72 | 72 | 4,384 |
  | Failing https://github.com/pytorch/pytorch/commit/3e263a46d03bbd64637b0607fe4d0d3c7ca0fa17 | 144 | 144 | 20,692 |

**Comment by rdspring1** [comment]
### FSDP-specific allocation pattern

Every layer's AllGather output is a fresh `torch.empty(unsharded_param_numel * world_size, ...)`. Lifecycle: alloc → AllGather → forward consumes → free. Every layer has an alloc and free call per step on the same `(NCCL_stream, unsharded_param_size)` bucket. This creates split/coalesce rhythm that #178362's per-block counter perturbs.
  
### Two notions of "deterministic"
1. Cross-rank determinism: same logical buffer chosen on every rank. Old address ordering didn't give this (per-process virtual addresses differ); #178362's per-block counter does by SPMD construction.
2. Step-over-step determinism: same address chosen on iteration N+1 as on iteration N. Old address ordering gave this as a side effect of stable segment placement; cudagraph trees relied on it.

#178362 gains (1) but breaks (2), because counters mint on every new Block(...), including every split-remainder. FSDP's per-step alloc/free pattern triggers heavy splitting in the equal-size cluster, perturbing counter order step-over-step.

#183591 fixes the conflict by minting counters at the segment level (rare, SPMD-aligned events) with address as the tiebreaker within each segment. Cross-rank determinism holds at the segment level, step-over-step determinism holds at the block level, and the two notions stop fighting each other.

-- Summarized by claude

**Comment by ngimel** [comment]
Fixed in #183591
